File: ios/OndaModule.swift
import Foundation
import OndaSDK
import React

/// RN 브리지 (iOS) — 무상태. invoke(method,args)를 OndaSDK 코어로 위임하고,
/// 코어 이벤트를 onda_pushOpened/received로 emit한다 (PRD-01A 4장).
/// 코어가 유일한 상태 보유자 — 이 클래스에는 큐/영속/토큰 상태가 없다.
@objc(OndaModule)
final class OndaModule: RCTEventEmitter {
  private var openedToken: UUID?
  private var receivedToken: UUID?

  override static func requiresMainQueueSetup() -> Bool { false }
  override func supportedEvents() -> [String]! { ["onda_pushOpened", "onda_pushReceived"] }

  override func startObserving() {
    openedToken = Onda.onPushOpened { [weak self] p in self?.forward("onda_pushOpened", p) }
    receivedToken = Onda.onPushReceived { [weak self] p in self?.forward("onda_pushReceived", p) }
  }

  override func stopObserving() {
    if let t = openedToken { Onda.off(t) }
    if let t = receivedToken { Onda.off(t) }
  }

  private func forward(_ event: String, _ p: PushPayload) {
    guard let json = OndaJSON.string(from: OndaJSON.dict(p)) else { return }
    sendEvent(withName: event, body: json)
  }

  /// 단일 dispatch 진입점 (PRD-01A 4장). argsJson → 코어 호출 → resultJson.
  @objc(invoke:argsJson:resolver:rejecter:)
  func invoke(_ method: String, argsJson: String,
              resolver resolve: @escaping RCTPromiseResolveBlock,
              rejecter reject: @escaping RCTPromiseRejectBlock) {
    let args = OndaJSON.dict(fromString: argsJson)
    switch method {
    case "initialize":
      guard let key = args["sdkKey"] as? String, let host = args["apiHost"] as? String,
            let url = URL(string: host) else { reject("E_ARGS", "initialize 인자 오류", nil); return }
      Onda.initialize(config: OndaConfig(sdkKey: key, apiHost: url))
      resolve(nil)
    case "identify":
      Onda.identify(externalId: args["externalId"] as? String ?? "")
      resolve(nil)
    case "reset": Onda.reset(); resolve(nil)
    case "setUserAttributes":
      Onda.setUserAttributes(OndaJSON.values(args["attrs"] as? [String: Any] ?? [:]))
      resolve(nil)
    case "track":
      Onda.track(args["name"] as? String ?? "", properties: args["properties"] as? [String: Any])
      resolve(nil)
    case "flush": Onda.flush(); resolve(nil)
    case "setPushSubscription":
      Onda.setPushSubscription(args["optedIn"] as? Bool ?? true); resolve(nil)
    case "setLogLevel": resolve(nil)
    case "getDeviceId": resolve(Onda.getDeviceId())
    case "getAnonId": resolve(Onda.getAnonId())
    case "getInitialPushPayload":
      resolve(Onda.getInitialPushPayload().flatMap { OndaJSON.string(from: OndaJSON.dict($0)) })
    case "replayBuffer":
      resolve(nil) // 코어 EventBus가 첫 구독 시 자동 재생 — no-op
    case "registerForPush":
      Task { let r = await Onda.registerForPush(); resolve(r.rawValue) }
    case "getPushSubscription":
      Task {
        let s = await Onda.getPushSubscription()
        resolve(OndaJSON.string(from: [
          "serviceOptIn": s.serviceOptIn,
          "osPermission": s.osPermission,
          "tokenRegistered": s.tokenRegistered,
        ]))
      }
    default:
      reject("E_METHOD", "알 수 없는 메서드: \(method)", nil)
    }
  }
}

/// 브리지 직렬화 헬퍼 (PRD-01A 4장 타입 규칙).
enum OndaJSON {
  static func dict(fromString s: String) -> [String: Any] {
    guard let data = s.data(using: .utf8),
          let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return [:] }
    return obj
  }
  static func string(from obj: Any) -> String? {
    guard let data = try? JSONSerialization.data(withJSONObject: obj) else { return nil }
    return String(data: data, encoding: .utf8)
  }
  static func dict(_ p: PushPayload) -> [String: Any] {
    var d: [String: Any] = ["messageId": p.messageId, "title": p.title, "body": p.body, "data": p.data]
    if let c = p.campaignId { d["campaignId"] = c }
    if let j = p.journeyId { d["journeyId"] = j }
    if let l = p.deepLink { d["deepLink"] = l }
    return d
  }
  static func values(_ raw: [String: Any]) -> [String: OndaValue] {
    raw.mapValues { v in
      switch v {
      case let s as String: return .string(s)
      case let b as Bool: return .bool(b)
      case let n as NSNumber: return .number(n.doubleValue)
      case let a as [String]: return .stringArray(a)
      default: return .null
      }
    }
  }
}

File: android/src/main/java/io/onda/rn/OndaModule.kt
package io.onda.rn

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.modules.core.DeviceEventManagerModule
import io.onda.sdk.Onda
import io.onda.sdk.OndaConfig
import io.onda.sdk.PushPayload
import org.json.JSONObject
import java.util.UUID

/**
 * RN 브리지 (Android) — 무상태. invoke(method,args)를 io.onda.sdk 코어로 위임하고,
 * 코어 이벤트를 onda_pushOpened/received로 emit한다 (PRD-01A 4장).
 */
class OndaModule(private val reactCtx: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactCtx) {

  private var openedToken: UUID? = null
  private var receivedToken: UUID? = null

  override fun getName() = "OndaModule"

  override fun initialize() {
    super.initialize()
    openedToken = Onda.onPushOpened { emit("onda_pushOpened", it) }
    receivedToken = Onda.onPushReceived { emit("onda_pushReceived", it) }
  }

  override fun invalidate() {
    openedToken?.let { Onda.off(it) }
    receivedToken?.let { Onda.off(it) }
    super.invalidate()
  }

  private fun emit(event: String, p: PushPayload) {
    reactCtx.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit(event, payloadJson(p))
  }

  /** 단일 dispatch 진입점. argsJson → 코어 호출 → resultJson(Promise). */
  @ReactMethod
  fun invoke(method: String, argsJson: String, promise: Promise) {
    val args = runCatching { JSONObject(argsJson) }.getOrDefault(JSONObject())
    when (method) {
      "initialize" -> {
        Onda.initialize(
          reactCtx.applicationContext,
          OndaConfig(sdkKey = args.optString("sdkKey"), apiHost = args.optString("apiHost")),
        )
        promise.resolve(null)
      }
      "identify" -> { Onda.identify(args.optString("externalId")); promise.resolve(null) }
      "reset" -> { Onda.reset(); promise.resolve(null) }
      "setUserAttributes" -> {
        Onda.setUserAttributes(args.optJSONObject("attrs").toMap()); promise.resolve(null)
      }
      "track" -> {
        Onda.track(args.optString("name"), args.optJSONObject("properties").toMap())
        promise.resolve(null)
      }
      "flush" -> { Onda.flush(); promise.resolve(null) }
      "setPushSubscription" -> { Onda.setPushSubscription(args.optBoolean("optedIn", true)); promise.resolve(null) }
      "setLogLevel" -> promise.resolve(null)
      "getDeviceId" -> promise.resolve(Onda.getDeviceId())
      "getAnonId" -> promise.resolve(Onda.getAnonId())
      "getInitialPushPayload" ->
        promise.resolve(Onda.getInitialPushPayload()?.let { payloadJson(it) })
      "replayBuffer" -> promise.resolve(null) // 코어 EventBus 자동 재생 — no-op
      "registerForPush" -> {
        // Activity 권한 요청은 호스트가 처리. 여기선 현재 상태 반영.
        Onda.registerForPush(currentActivity) { r -> promise.resolve(r.name.lowercase()) }
      }
      "getPushSubscription" -> {
        val s = Onda.getPushSubscription()
        promise.resolve(
          JSONObject()
            .put("serviceOptIn", s.serviceOptIn)
            .put("osPermission", s.osPermission)
            .put("tokenRegistered", s.tokenRegistered)
            .toString(),
        )
      }
      else -> promise.reject("E_METHOD", "알 수 없는 메서드: $method")
    }
  }

  // RN NativeEventEmitter 요건 (no-op — DeviceEventEmitter 사용).
  @ReactMethod fun addListener(eventName: String) {}
  @ReactMethod fun removeListeners(count: Int) {}

  private fun payloadJson(p: PushPayload): String = JSONObject().apply {
    put("messageId", p.messageId)
    p.campaignId?.let { put("campaignId", it) }
    p.journeyId?.let { put("journeyId", it) }
    put("title", p.title)
    put("body", p.body)
    p.deepLink?.let { put("deepLink", it) }
    put("data", JSONObject(p.data as Map<*, *>))
  }.toString()

  private fun JSONObject?.toMap(): Map<String, Any?> {
    if (this == null) return emptyMap()
    return keys().asSequence().associateWith { k -> if (isNull(k)) null else get(k) }
  }
}
